### PR TITLE
Add CIE Clear Sky model to procedural environment

### PR DESCRIPTION
**Pull request by OmarShehata**
This is the same as https://github.com/AnalyticalGraphicsInc/cesium/pull/7123 but re-branched off the base branch.

You should be able to actually test this out here @bagnell .

I'm going to experiment with multiplying by the atmosphere color to get an orange-ish tint as the sun sets relative to the object. Will bump this PR with some pictures once that's in.

### CODE REVIEW

**Comment by OmarShehata** [comment]
I've tried adding the atmosphere color using the `GroundAtmosphere` shader, and it does look nice as it goes from yellow to red as the sun sets:

sunset

However, if you look real close, there are these noisy artifacts, I'm not sure if that's just because the ground atmosphere wasn't intended to be used at this scale @bagnell ?

look_close

The color also seems to be determined by the camera direction, but it seems to me that it should be purely based on the angle from the sun to the horizon.

I think it would be simpler and would look better simply to use that computed angle to add a reddish tint for small values of that angle. I'm still tweaking it to find a model that looks good yet subtle.

**Comment by OmarShehata** [comment]
@bagnell I think this should be ready to merge. I couldn't find an easy way to make a reddish tint at sunrise/sunset look good.

File: Source/Scene/processPbrMaterials.js [hunk 1]
```diff
@@ -688,10 +688,18 @@ define([
             fragmentShader += '    specularIrradiance = mix(specularIrradiance, belowHorizonColor, smoothstep(aroundHorizon, farBelowHorizon, reflectionDotNadir) * inverseRoughness);\n';
             fragmentShader += '    specularIrradiance = mix(specularIrradiance, nadirColor, smoothstep(farBelowHorizon, 1.0, reflectionDotNadir) * inverseRoughness);\n';
 
+            // Luminance model from page 40 of http://silviojemma.com/public/papers/lighting/spherical-harmonic-lighting.pdf
+            // Angle between sun and zenith
+            fragmentShader += '    float S = acos(clamp(dot(normalize(czm_inverseViewRotation * l), normalize(v_positionWC * -1.0)), 0.001, 1.0));\n';
+            // Angle between zenith and current pixel
+            fragmentShader += '    float theta = acos(clamp(dot(normalize(czm_inverseViewRotation * n), normalize(v_positionWC * -1.0)), 0.001, 1.0));\n';
+            // Angle between sun and current pixel
+            fragmentShader += '    float gamma = acos(NdotL);\n';
```
Comment (bagnell) [inline Source/Scene/processPbrMaterials.js]: For `S`, `theta`, and `gamma`, you don't need to use `acos` just to call ` cos` on it again. For example, in the luminance below, there is `pow(cos(gamma), 2.0)` which is `pow(cos(acos(NdotL)), 2.0` which can be simplified to just `NdotL * NdotL`. You can do that for all three.

File: Source/Scene/processPbrMaterials.js [hunk 2]
```diff
@@ -688,10 +688,18 @@ define([
...
+            fragmentShader += '    float luminanceAtZenith = 10.0;\n';
+            fragmentShader += '    float luminance = luminanceAtZenith * (((0.91 + 10.0 * exp(-3.0 * gamma) + 0.45 * pow(cos(gamma), 2.0)) * (1.0 - exp(-0.32 / cos(theta)))) / (0.91 + 10.0 * exp(-3.0 * S) + 0.45 * pow(cos(S),2.0)) * (1.0 - exp(-0.32)));\n';
```
Comment (bagnell) [inline Source/Scene/processPbrMaterials.js]: Can you split this up for readability? Perhaps, `numerator` and `denominator` and `luminance = luminanceAtZenith * (numerator/ denominator);`

File: Source/Scene/processPbrMaterials.js [hunk 3]
```diff
@@ -688,10 +688,18 @@ define([
...
             fragmentShader += '    specularIrradiance = mix(specularIrradiance, nadirColor, smoothstep(farBelowHorizon, 1.0, reflectionDotNadir) * inverseRoughness);\n';
 
+            // Luminance model from page 40 of http://silviojemma.com/public/papers/lighting/spherical-harmonic-lighting.pdf
+            // Angle between sun and zenith
+            fragmentShader += '    float S = acos(clamp(dot(normalize(czm_inverseViewRotation * l), normalize(v_positionWC * -1.0)), 0.001, 1.0));\n';
+            // Angle between zenith and current pixel
+            fragmentShader += '    float theta = acos(clamp(dot(normalize(czm_inverseViewRotation * n), normalize(v_positionWC * -1.0)), 0.001, 1.0));\n';
+            // Angle between sun and current pixel
+            fragmentShader += '    float gamma = acos(NdotL);\n';
+            fragmentShader += '    float luminanceAtZenith = 10.0;\n';
```
Comment (bagnell) [inline Source/Scene/processPbrMaterials.js]: Where did you get this value? Should it be configurable and a uniform?

**Comment by bagnell** [comment]
I think it looks good. What do you think @emackey?
With the luminance:
image
Without the luminance:
image

**Comment by OmarShehata** [comment]
@bagnell good suggestion about splitting it into numerator and denominator, I caught a missing paren. Subsequently, the new luminanceAtZenith value that looks good is 1.0 instead of 10.0. I changed it to a uniform and exposed it as a property on the model.

I also used the dot product directly where appropriate instead of `cos(acos(..))`. Should be good now.

**Comment by lilleyse** [comment]
@OmarShehata @bagnell can this be merged into `ibl`?

**Comment by OmarShehata** [comment]
It's good on my end @lilleyse. The only thing Dan was wondering about was whether the default luminance of 10 was too much since I only tested it on the spheres, but I just tried with other models and I think it's a good default (and I imagine it would look a lot better in the HDR branch). It's easy to tweak the default later in any case.

**Comment by lilleyse** [comment]
Ok, it should be fine to merge now then and review everything together in https://github.com/AnalyticalGraphicsInc/cesium/pull/7172.

**Pull request merged**